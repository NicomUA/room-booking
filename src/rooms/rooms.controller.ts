import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  ParseIntPipe,
  Post,
  UseGuards,
} from '@nestjs/common';
import { RoomsService } from './rooms.service';
import { Room } from '@prisma/client';
import { CreateRoomDto } from './dto/create-room.dto';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { AuthGuard } from '../auth/auth.guard';

@ApiTags('rooms')
@Controller('rooms')
@ApiBearerAuth()
@UseGuards(AuthGuard)
export class RoomsController {
  constructor(private readonly roomsService: RoomsService) {}

  @Get()
  async findAll(): Promise<Room[]> {
    return this.roomsService.findAll();
  }

  @Get(':id')
  async findById(@Param('id', ParseIntPipe) id: number): Promise<Room> {
    return this.roomsService.findById(id);
  }

  @Post()
  async create(@Body() data: CreateRoomDto): Promise<Room> {
    return this.roomsService.create(data);
  }

  @Delete(':id')
  async delete(@Param('id', ParseIntPipe) id: number): Promise<void> {
    return this.roomsService.delete(id);
  }
}
