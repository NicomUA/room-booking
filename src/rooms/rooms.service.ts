// room.service.ts
import { Injectable } from '@nestjs/common';
import { Room, Prisma } from '@prisma/client';
import { DbService } from '@app/db';

@Injectable()
export class RoomsService {
  constructor(private db: DbService) {}

  async findAll(): Promise<Room[]> {
    return this.db.room.findMany();
  }

  async findById(id: number): Promise<Room> {
    return this.db.room.findUnique({
      where: { id },
    });
  }

  async create(data: Prisma.RoomCreateInput): Promise<Room> {
    return this.db.room.create({ data });
  }

  async update(id: number, data: Prisma.RoomUpdateInput): Promise<Room> {
    return this.db.room.update({
      where: { id },
      data,
    });
  }

  async delete(id: number): Promise<void> {
    await this.db.room.delete({
      where: { id },
    });
  }
}
