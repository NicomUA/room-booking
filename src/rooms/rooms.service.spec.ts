import { RoomsService } from './rooms.service';
import { DbService } from '@app/db';
import { TestBed } from '@automock/jest';

describe('RoomsService', () => {
  let service: RoomsService;
  let dbService: jest.Mocked<DbService>;

  const findManyStub = jest.fn().mockResolvedValue([]);
  const findUniqueStub = jest.fn().mockResolvedValue({
    id: 1,
    name: 'test',
  });
  const createStab = jest.fn();
  const deleteStab = jest.fn();

  beforeAll(() => {
    const { unit, unitRef } = TestBed.create(RoomsService).compile();
    service = unit;
    dbService = unitRef.get(DbService);

    dbService.room.findMany = findManyStub;
    dbService.room.findUnique = findUniqueStub;
    dbService.room.create = createStab;
    dbService.room.delete = deleteStab;
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('findAll', async () => {
    findManyStub.mockResolvedValue([]);
    const result = await service.findAll();
    expect(result).toEqual([]);
  });

  describe('findById', () => {
    it('when room not found', async () => {
      findUniqueStub.mockResolvedValue(undefined);
      const response = async () => await service.findById(1);
      expect(response).rejects.toThrowErrorMatchingSnapshot();
    });

    it('when room found', async () => {
      findUniqueStub.mockResolvedValue({
        id: 1,
        name: 'test',
      });
      const result = await service.findById(1);
      expect(result).toEqual({
        id: 1,
        name: 'test',
      });
    });
  });

  it('create', async () => {
    createStab.mockResolvedValue({
      id: 1,
      name: 'test',
    });
    const result = await service.create({
      name: 'test',
    });
    expect(result).toEqual({
      id: 1,
      name: 'test',
    });
  });

  describe('delete', () => {
    it('when room not found', async () => {
      findUniqueStub.mockResolvedValue(undefined);
      const response = async () => await service.delete(1);
      expect(response).rejects.toThrowErrorMatchingSnapshot();
    });

    it('when room found', async () => {
      deleteStab.mockResolvedValue(undefined);
      findUniqueStub.mockResolvedValue({
        id: 1,
        name: 'test',
      });
      const result = await service.delete(1);
      expect(result).toBeUndefined();
    });
  });
});
