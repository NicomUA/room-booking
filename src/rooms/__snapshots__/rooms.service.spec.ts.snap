// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomsService delete when room not found 1`] = `"Room not found"`;

exports[`RoomsService findById when room not found 1`] = `"Room not found"`;
